extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -134,46 +134,37 @@ public Optional<InputStream> streamTaskLog(String taskid, long offset)
   @Override
   public ListenableFuture<TaskStatus> run(Task task)
   {
-    return tasks.computeIfAbsent(
-        task.getId(), k -> new KubernetesWorkItem(task, exec.submit(() -> runTask(task)))
-    ).getResult();
+    return runOrJoinTask(task, true);
   }
 
-  protected ListenableFuture<TaskStatus> joinAsync(Task task)
+  protected ListenableFuture<TaskStatus> runOrJoinTask(Task task, boolean run)
   {
-    return tasks.computeIfAbsent(
-        task.getId(), k -> new KubernetesWorkItem(task, exec.submit(() -> joinTask(task)))
-    ).getResult();
-  }
-
-  private TaskStatus runTask(Task task)
-  {
-    return doTask(task, true);
-  }
-
-  private TaskStatus joinTask(Task task)
-  {
-    return doTask(task, false);
+    synchronized (tasks) {
+      tasks.computeIfAbsent(task.getId(), k -> new KubernetesWorkItem(task, exec.submit(() -> doTask(task, run))));
+      return tasks.get(task.getId()).getResult();
+    }
   }
 
   @VisibleForTesting
   protected TaskStatus doTask(Task task, boolean run)
   {
-    KubernetesPeonLifecycle peonLifecycle = peonLifecycleFactory.build(task);
+    try {
+      KubernetesPeonLifecycle peonLifecycle = peonLifecycleFactory.build(task);
 
-    KubernetesWorkItem workItem = tasks.get(task.getId());
+      synchronized (tasks) {
+        KubernetesWorkItem workItem = tasks.get(task.getId());
 
-    if (workItem == null) {
-      throw new ISE("Task [%s] disappeared", task.getId());
-    }
+        if (workItem == null) {
+          throw new ISE("Task [%s] disappeared", task.getId());
```

Comment:
Should we maybe just return a failed `TaskStatus` here instead of throwing an exception? The exception thrown by this method may or may not be handled by the calling code, but no point depending on that if we already know that the reason for the task failure.

We should do the same thing in the catch block too.

But this doesn't need to be done as a part of this PR, just wanted to call it out.

Reply:
Makes sense, we can revisit this later.